File: ds/graph/lc_934_shortest_bridge/java/Solution.java
import java.util.*;
class Solution {
    static int[] dx = {1, -1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    public int ShortestBridge(int[][] A) {
        if (A == null || A.length == 0 || A[0].length == 0) return 0;

        int m = A.length, n = A[0].length;
        Queue<int[]> queue = new LinkedList<>();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (A[i][j] == 1) {
                    dfs(A, i, j, queue);
                    break;
                }
            }
            if (!queue.isEmpty()) {
                break;
            }
        }

        int step = 0;
        while (!queue.isEmpty()) {
            int k = queue.size();
            while (k-- > 0) {
                int[] p = queue.poll();
                for (int i = 0; i < 4; i++) {
                    int x = p[0] + dx[i];
                    int y = p[1] + dy[i];
                    if (x >= 0 && x < m && y >= 0 && y < n) {
                        if (A[x][y] == 1) {
                            return step;
                        }else if (A[x][y] == 0) {
                            A[x][y] = -1;
                            queue.offer(new int[]{x, y});
                        }
                    }
                }
            }
            step++;
        }
        return -1;
    }

    void dfs(int[][] A, int i, int j, Queue<int[]> queue) {
        A[i][j] = -1;
        queue.offer(new int[]{i, j});
        for (int k = 0; k < 4; k++) {
            int x = i + dx[k];
            int y = j + dy[k];
            if (x >= 0 && x < A.length && y >= 0 && y < A[0].length && A[x][y] == 1) {
                dfs(A, x, y, queue);
            }
        }
    }
}

File: ds/graph/lc_934_shortest_bridge/python/Solution.py
class Solution:
    def shortestBridge(self, A):

        def dfs(i, j):
            A[i][j] = -1
            queue.append((i, j))
            for x, y in ((i-1, j), (i+1, j), (i, j-1), (i, j+1)):
                if 0 <= x < n and 0 <= y < n and A[x][y] == 1:
                    dfs(x, y)

        def first():
            for i in range(n):
                for j in range(n):
                    if A[i][j]:
                        return i, j;

        n, step, queue = len(A), 0, []

        dfs(*first())
        while queue:
            new = []
            for i, j in queue:
                for x, y in ((i-1, j), (i+1, j), (i, j-1), (i, j+1)):
                    if 0 <= x < n and 0 <= y < n:
                        if A[x][y] == 1:
                            return step
                        elif A[x][y] == 0:
                            A[x][y] = -1
                            new.append((x, y))
            step += 1
            bfs = new